hack/generator/pkg/codegen/pipeline/create_conversion_graph.go

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT license.
+ */
+
+package pipeline
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+
+	"github.com/Azure/azure-service-operator/hack/generator/pkg/astmodel"
+	"github.com/Azure/azure-service-operator/hack/generator/pkg/codegen/storage"
+)
+
+// CreateConversionGraphStageId is the unique identifier for this stage
+const CreateConversionGraphStageId = "createConversionGraph"
+
+// CreateConversionGraph walks the set of available types and creates a graph of conversions that will be used to
+// convert resources to/from the designated storage (or hub) version
+func CreateConversionGraph(conversionGraph *storage.ConversionGraph) Stage {
+	stage := MakeStage(
+		CreateConversionGraphStageId,
+		"Create the graph of conversions between versions of each resource group",
+		func(ctx context.Context, types astmodel.Types) (astmodel.Types, error) {
+			// Collect all distinct references
+			allReferences := astmodel.NewPackageReferenceSet()
+			for _, def := range types {
+				allReferences.AddReference(def.Name().PackageReference)
+			}
+
+			// For each one, create a storage reference to match
+			sortedApiReferences := allReferences.AsSlice()
+			sortedStorageReferences := make([]astmodel.PackageReference, len(sortedApiReferences))
+			astmodel.SortPackageReferencesByPathAndVersion(sortedApiReferences)
+			for index, ref := range sortedApiReferences {
+				localRef, ok := ref.AsLocalPackage()
+				if !ok {
+					// Shouldn't have any non-local references, if we do, abort
+					return nil, errors.Errorf("expected all API references to be local references, but %s was not", ref)
+				}
+
+				storageRef := astmodel.MakeStoragePackageReference(localRef)
+				sortedStorageReferences[index] = storageRef
+				conversionGraph.AddLink(localRef, storageRef)
+			}
+
+			// For each Preview API version, the link goes from the associated storage version to the immediately prior
+			// storage version, no matter whether it's preview or GA.
+			// We use the API reference to work out whether it's preview or not, because the suffix used for storage
+			// versions means IsPreview() always returns true
+			for i, ref := range sortedApiReferences {
+				if i == 0 || !ref.IsPreview() {
+					continue
+				}
+
+				conversionGraph.AddLink(sortedStorageReferences[i], sortedStorageReferences[i-1])
+			}
```

Comment:
I don't understand this part. Don't the references need to be split by group before you can do this? Otherwise you'll get the earliest version of one group pointing at the latest version of the previous group - something like `microsoft.network/v20150101` pointing to `microsoft.compute/v20210701`. In my head each stage runs with the full set of loaded types. Is that right? Or is there something structural that means that the stage only runs with (all versions of) the types from one group?

Reply:
Yes, you're perfectly correct - I transplated this code from #1533 and put it in the wrong place. 😲 

I'll relocate it into `GroupConversionGraph` and add some tests to ensure it behaves correctly.